New line characters are removed from stack traces in testng-results.xml

### TestNG Version
6.13.1

### Expected behavior
Stack trace in testng-results.xml contains new line characters.

### Actual behavior
New line characters are removed from stack traces.

### Is the issue reproducible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
```
    @Test
    public void test()
    {
        Assert.fail( "failed" );
    }
```

Results in 6.12
```
            <full-stacktrace>
              <![CDATA[java.lang.AssertionError: failed
	at Tester.test(Tester.java:12)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:498)
]]>
            </full-stacktrace>
```

Results in 6.13
```
            <full-stacktrace>
              <![CDATA[java.lang.AssertionError: failedat Tester.test(Tester.java:12)at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)at java.lang.reflect.Method.invoke(Method.java:498)]]>
            </full-stacktrace>
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,5 +1,6 @@
 Current
 =======
+Fixed: GITHUB-1659: New line characters are removed from stack traces in testng-results.xml (Krishnan Mahadevan)
 Fixed: GITHUB-1503: Consider adding a -n (no execute) option (Krishnan Mahadevan)
 Fixed: GITHUB-1648: Depends on method is not respected on the sequential run on second test that extends same base testClass (Krishnan Mahadevan)
 Fixed: GITHUB-1636: Parallel test run is not working in 6.13.1 (Krishnan Mahadevan)

--- a/src/main/java/org/testng/reporters/XMLStringBuffer.java
+++ b/src/main/java/org/testng/reporters/XMLStringBuffer.java
@@ -306,7 +306,8 @@ public class XMLStringBuffer {
    */
   public void addCDATA(String content) {
     if (content != null) {
-        content = content.replaceAll("\\p{Cc}", "");
+        //Solution from https://coderanch.com/t/455930/java/Remove-control-characters
+        content = content.replaceAll("[\\p{Cc}&&[^\\r\\n]]", "");
     }
     m_buffer.append(m_currentIndent);
     if (content == null) {

--- a/src/test/java/test/reports/Issue1659Sample.java
+++ b/src/test/java/test/reports/Issue1659Sample.java
@@ -1,0 +1,14 @@
+package test.reports;
+
+import org.testng.annotations.Test;
+
+public class Issue1659Sample {
+    @Test
+    public void testMethod() {
+        triggerExceptions();
+    }
+
+    private void triggerExceptions() {
+        throw new RuntimeException("simulating test failure");
+    }
+}

--- a/src/test/java/test/reports/XmlReporterTest.java
+++ b/src/test/java/test/reports/XmlReporterTest.java
@@ -1,11 +1,16 @@
 package test.reports;
 
-import org.testng.Assert;
 import org.testng.TestNG;
 import org.testng.annotations.Test;
 import org.testng.reporters.XMLReporter;
+import org.w3c.dom.Document;
 import test.SimpleBaseTest;
 
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathFactory;
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileReader;
@@ -13,18 +18,15 @@ import java.io.IOException;
 import java.util.UUID;
 import java.util.regex.Pattern;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 public class XmlReporterTest extends SimpleBaseTest {
     @Test(description = "GITHUB-1566")
     public void testMethod() throws IOException {
-        String suiteName = UUID.randomUUID().toString();
-        File fileLocation = createDirInTempDir(suiteName);
-        TestNG testng = create(fileLocation.toPath(), Issue1566Sample.class);
-        testng.setUseDefaultListeners(true);
-        testng.run();
-        File file = new File(fileLocation, XMLReporter.FILE_NAME);
+        File file = runTest(Issue1566Sample.class);
         boolean flag = false;
         Pattern pattern = Pattern.compile("\\p{Cc}");
-        try(BufferedReader reader = new BufferedReader(new FileReader(file))) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
             String line;
             while ((line = reader.readLine()) != null) {
                 if (pattern.matcher(line).find()) {
@@ -32,6 +34,30 @@ public class XmlReporterTest extends SimpleBaseTest {
                 }
             }
         }
-        Assert.assertFalse(flag, "Should not have found a control character");
+        assertThat(flag).as("Should not have found a control character").isFalse();
+    }
+
+    @Test(description = "GITHUB1659")
+    public void ensureStackTraceHasLineFeedsTest() throws Exception {
+        File file = runTest(Issue1659Sample.class);
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder builder = factory.newDocumentBuilder();
+        Document doc = builder.parse(file);
+        XPath xPath = XPathFactory.newInstance().newXPath();
+        String expression = "//full-stacktrace/text()";
+        String data = (String) xPath.compile(expression).evaluate(doc, XPathConstants.STRING);
+        data = data.trim();
+        assertThat(data.split("\n").length)
+                .as("Line feeds and carriage returns should not have been removed")
+                .isGreaterThan(1);
+    }
+
+    private static File runTest(Class<?> clazz) {
+        String suiteName = UUID.randomUUID().toString();
+        File fileLocation = createDirInTempDir(suiteName);
+        TestNG testng = create(fileLocation.toPath(), clazz);
+        testng.setUseDefaultListeners(true);
+        testng.run();
+        return new File(fileLocation, XMLReporter.FILE_NAME);
     }
 }

